File: dot-config/waybar/scripts/adapters/system_time.py
"""Wall-clock ``TimePort`` adapter."""

from __future__ import annotations

import time
from datetime import datetime


class SystemTime:
    def now_ms(self) -> int:
        return int(time.time() * 1000)

    def elapsed_ms(self, start_ms: int) -> int:
        return self.now_ms() - start_ms

    def now(self) -> datetime:
        return datetime.now()

File: dot-config/hypr/adapters/system_time.lua
-- Adapter: TimePort -> wall-clock process timer.
-- Uses Linux /proc/uptime (monotonic, millisecond precision); falls back to
-- os.time() on other platforms. os.clock() is deliberately avoided: it reports
-- CPU time, which under-reports elapsed wall time.

local M = {}
M.__index = M

local UPTIME_PATH = "/proc/uptime"

function M.new()
  return setmetatable({}, M)
end

local function read_uptime_ms()
  local file = io.open(UPTIME_PATH, "r")
  if not file then
    return nil
  end
  local line = file:read("*l")
  file:close()
  if not line then
    return nil
  end
  local seconds = tonumber(line:match("^(%d+%.?%d*)"))
  if not seconds then
    return nil
  end
  return math.floor(seconds * 1000)
end

function M:now_ms()
  return read_uptime_ms() or (os.time() * 1000)
end

function M:elapsed_ms(start_ms)
  return self:now_ms() - start_ms
end

return M
